Update ValidityState docs (#26431)
mdn/content: files/en-us/web/api/validitystate/customerror/index.md

---
title: "ValidityState: customError property"
short-title: customError
slug: Web/API/ValidityState/customError
page-type: web-api-instance-property
browser-compat: api.ValidityState.customError
---

{{APIRef("HTML DOM")}}

The read-only **`customError`** property of the [`ValidityState`](/en-US/docs/Web/API/ValidityState) interface if the element's custom validity message has been set to a non-empty string by calling the element's {{domxref("HTMLObjectElement.setCustomValidity", "setCustomValidity()")}} method.

## Value

A boolean that is `true` if a custom error message has been set to a non-empty string.

## Examples

### Detecting a custom error

> In the following example, we set a custom error when both emails don't match. A log is written when this is the case, and a different message is displayed when there is no custom errors. When there is an error, custom or not, the widget are surrounded by a red border.

Author: The example is a bit artificial, as the function detects the error only to log a message. I've not found an idea of a simple example that would be better.

#### HTML

```html

> <input type="email" id="original" /> <input type="email" id="confirmation" />

Author: Done in [5ff1189](https://github.com/mdn/content/pull/26431/commits/5ff11893345e9237dfcac5378b94dac5020f5242)

<pre id="log"></pre>
```

#### CSS

```css
input:invalid {
  border: red solid 3px;
}
```

#### JavaScript

```js
const log = document.getElementById("log");
const confirmElement = document.getElementById("confirmation");

confirmElement.addEventListener("change", () => {
  const originalElement = document.getElementById("original");
  confirmElement.setCustomValidity(
    confirmElement.value != originalElement.value
      ? "Both email addresses must match."
      : ""
  );
});

function displayLog() {
  if (confirmElement.validity.customError) {
    log.textContent = "A custom error message has been detected.";
  } else {
    log.textContent = "No custom error message is set.";
  }
  requestAnimationFrame(displayLog);
}

> requestAnimationFrame(displayLog);

Author: The `displayLog` function logs the message. It needs to be called periodically (here every frame, but we could do it less often) to update the log.

I didn't want to put it into the event handler because the event handler sets the message so, it doesn't need to read `customError` to know one has been set.

I've not found a simple example that would be less artificial.

---
title: "ValidityState: customError property"
short-title: customError
slug: Web/API/ValidityState/customError
page-type: web-api-instance-property
browser-compat: api.ValidityState.customError
---

{{APIRef("HTML DOM")}}

The read-only **`customError`** property of the [`ValidityState`](/en-US/docs/Web/API/ValidityState) interface returns `true` if an element doesn't meet the validation required in the custom validity set by the element's {{domxref("HTMLObjectElement.setCustomValidity", "setCustomValidity()")}} method.

## Value

A boolean that is `true` if a custom error message has been set to a non-empty string.

## Examples

### Detecting a custom error

> In the following example, we set a custom error when both emails don't match. A log is written when this is the case, and a different message is displayed when there is no custom errors. When there is an error, custom or not, the widget is surrounded by a red border.

Reviewer: If you think the suggestion below is good, we could say something like:

>In the example below, `setCustomValidity` sets a custom error message for a field when the input is invalid. If you click the "Check input" button, we call `reportValidity`, which displays a validation message if the input value doesn't match the constraints. If you enter the text "good" and check the form validity, the field is marked invalid until the custom error message is cleared (set to an empty string).

Reviewer: Linking to https://developer.mozilla.org/en-US/docs/Web/HTML/Constraint_validation#constraint_validation_process maybe

> In the following example, we set a custom error when both emails don't match. A log is written when this is the case, and a different message is displayed when there is no custom errors. When there is an error, custom or not, the widget is surrounded by a red border.

Reviewer: ```suggestion
In the example below, `setCustomValidity` sets a custom error message for a field when the input is invalid. If you click the "Check input" button, we call `reportValidity`, which displays a validation message if the input value doesn't match the constraints. If you enter the text "good" and check the form validity, the field is marked invalid until the custom error message is cleared (set to an empty string). 
```

#### HTML

```html
<label for="original">Type your email:</label>
<input type="email" id="original" />
<label for="confirmation">Retype your email:</label>
<input type="email" id="confirmation" />

> <pre id="log"></pre>

Reviewer: What about the [example from the spec](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#the-constraint-validation-api:custom-validity-error-message):

https://codepen.io/bsmth/pen/abrYEJd

> <label for="original">Type your email:</label>
> <input type="email" id="original" />
> <label for="confirmation">Retype your email:</label>
> <input type="email" id="confirmation" />
> <pre id="log"></pre>

Reviewer: ```suggestion
<input type="text" id="feelingInput" placeholder="How are you feeling?" oninput="check(this)">
<button onclick="validateInput()">Check input</button>
```

Reviewer 2: I didn't think we use onEvent attributes on MDN. Do we?

Reviewer: Good call, fixed in 94ee94117fd1df4f761aa36ba7029919275dce43

```

#### CSS

```css
input:invalid {
  border: red solid 3px;
}
```

#### JavaScript

```js

> const log = document.getElementById("log");
> const confirmElement = document.getElementById("confirmation");
> 
> confirmElement.addEventListener("change", () => {
>   const originalElement = document.getElementById("original");
>   confirmElement.setCustomValidity(
>     confirmElement.value != originalElement.value
>       ? "Both email addresses must match."
>       : "",
>   );
> });
> 
> function displayLog() {
>   if (confirmElement.validity.customError) {
>     log.textContent = `There is a custom error on the repeated password: ${confirmElement.validationMessage}`;
>   } else {
>     log.textContent = "No custom error message is set.";
>   }
>   requestAnimationFrame(displayLog);
> }
> 
> requestAnimationFrame(displayLog);

Reviewer: ```suggestion
function check(input) {
  if (input.value == "good") {
    input.setCustomValidity('"' + input.value + '" is not a feeling.');
  } else {
    // Reset the custom error message
    input.setCustomValidity("");
  }
}

function validateInput() {
  document.getElementById("feelingInput").reportValidity();
}
```
